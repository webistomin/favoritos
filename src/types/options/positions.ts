export enum IFavoritosPositions {
  TOP_LEFT = 'top-left',
  TOP_RIGHT = 'top-right',
  BOTTOM_LEFT = 'bottom-left',
  BOTTOM_RIGHT = 'bottom-right',
}
