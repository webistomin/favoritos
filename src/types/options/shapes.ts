export enum IFavoritosShape {
  CIRCLE = 'circle',
  RECT = 'rectangle',
}
