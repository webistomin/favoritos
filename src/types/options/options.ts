import { IFavoritosPositions } from './positions';
import { IFavoritosShapes } from './shapes';

export interface IFavoritosOption {
  icon?: {
    iconSelector?: string;
    backgroundColor?: string | string[];
    shape?: IFavoritosShapes;
    lineWidth?: number;
    width?: number;
    height?: number;
  };
  badge?: {
    fontSize?: number;
    fontFamily?: string;
    backgroundColor?: string | string[];
    color?: string;
    position?: IFavoritosPositions;
    shape?: IFavoritosShapes;
    minWidth?: number;
    minHeight?: number;
  };
  debug?: {
    enabled?: boolean;
    debugSelector?: string;
  };
}
