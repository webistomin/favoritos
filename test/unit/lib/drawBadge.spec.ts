import Favoritos from '../../../src/favoritos';
import { IFavoritosShape } from '../../../src/types/options/shapes';

describe('Favoritos: drawBadge method', () => {
  beforeEach(() => {
    document.head.innerHTML = `<link rel="shortcut icon" href="#">`;
  });

  it('draw badge without count param', () => {
    const lib = new Favoritos();
    expect(() => lib['drawBadge']()).not.toThrow(TypeError);
    expect(() => lib['drawBadge']()).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('draw badge if params are correct', () => {
    const lib = new Favoritos();
    expect(() => lib['drawBadge'](50)).not.toThrow(TypeError);
    expect(() => lib['drawBadge'](50)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('correctly draw badge with favicon from cache', () => {
    const lib = new Favoritos();
    const image = new Image();
    lib['userIconCache'] = image;
    expect(() => lib['drawBadge'](50)).not.toThrow(TypeError);
    expect(() => lib['drawBadge'](50)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('draw badge with circle shape', () => {
    const lib = new Favoritos({
      badge: {
        shape: IFavoritosShape.CIRCLE,
      },
    });
    expect(() => lib['drawBadge'](50)).not.toThrow(TypeError);
    expect(() => lib['drawBadge'](50)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });

  it('draw badge with rect shape', () => {
    const lib = new Favoritos({
      badge: {
        shape: IFavoritosShape.RECT,
      },
    });
    expect(() => lib['drawBadge'](25)).not.toThrow(TypeError);
    expect(() => lib['drawBadge'](25)).not.toThrow(DOMException);
    // @ts-ignore
    expect(lib['iconCanvasContext'].__getEvents()).toMatchSnapshot();
  });
});
