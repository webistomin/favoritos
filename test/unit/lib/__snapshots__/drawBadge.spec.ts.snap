// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favoritos: drawBadge method correctly draw badge with favicon from cache 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "text": "50",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "height": 32,
      "width": 32,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "value": "#d21f3c",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#d21f3c",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineJoin",
  },
  Object {
    "props": Object {
      "value": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "height": 12,
      "width": 12,
      "x": 15,
      "y": 15,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 12,
      "width": 12,
      "x": 15,
      "y": 15,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#fff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "maxWidth": 32,
      "text": "50",
      "x": 21,
      "y": 22.53,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "closePath",
  },
  Object {
    "props": Object {
      "text": "50",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "measureText",
  },
  Object {
    "props": Object {
      "height": 32,
      "width": 32,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "value": "#d21f3c",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "value": "#d21f3c",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": "round",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineJoin",
  },
  Object {
    "props": Object {
      "value": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "height": 12,
      "width": 12,
      "x": 15,
      "y": 15,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeRect",
  },
  Object {
    "props": Object {
      "height": 12,
      "width": 12,
      "x": 15,
      "y": 15,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "value": "#fff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "maxWidth": 32,
      "text": "50",
      "x": 21,
      "y": 22.53,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "closePath",
  },
]
`;

exports[`Favoritos: drawBadge method draw badge if params are correct 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: drawBadge method draw badge with circle shape 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: drawBadge method draw badge with rect shape 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;

exports[`Favoritos: drawBadge method draw badge without count param 1`] = `
Array [
  Object {
    "props": Object {
      "value": "18px Helvetica, Arial, sans-serif",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "middle",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textBaseline",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
]
`;
